src/main/scala/fm/xml/XmlReader.scala

```diff
@@ -64,143 +101,186 @@ object XmlReader {
+  if (overrideDefaultNamespaceURI.isNotBlank) require(defaultNamespaceURI.isBlank, "Can't set both defaultNamespaceURI and overrideDefaultNamespaceURI")
+
   private def wrapXMLStreamReader2(r: XMLStreamReader2): XMLStreamReader2 = {
     if (overrideDefaultNamespaceURI.isNotBlank) new XmlReader.OverrideDefaultNamespaceStreamReaderDelegate(r, overrideDefaultNamespaceURI)
     else if (defaultNamespaceURI.isNotBlank) new XmlReader.DefaultNamespaceStreamReaderDelegate(r, defaultNamespaceURI)
     else r
   }
-  
+
   protected def foreachWithResource[U](f: T => U, reader: Reader) {
-    val unmarshaller: Unmarshaller = jaxbContext.createUnmarshaller()
+
     val inputFactory = new WstxInputFactory()
     inputFactory.setProperty(XMLInputFactory.SUPPORT_DTD, false)
     inputFactory.configureForSpeed()
-    
+
     import Resource.toCloseable
-    
+
     Resource.using(wrapXMLStreamReader2(inputFactory.createXMLStreamReader(XmlInvalidCharFilter(reader)).asInstanceOf[XMLStreamReader2])) { xmlStreamReader: XMLStreamReader2 =>
-      // The current depth we are in the tree 
+      // The current depth we are in the tree
       // (which should correspond with the itemPath index we are currently interested in)
       var currentDepth: Int = 0
+      var currentPath: Array[String] = Array.empty
+      var done: Boolean = false
 
       // Move to the ROOT element (skipping stuff like DTDs) and check it's name and then move to the next parsing event
       xmlStreamReader.seekFirstEventPastRootElement(rootName)
-      
-      def seekToNextElement(): Unit = currentDepth = seekToNextElementImpl(xmlStreamReader, currentDepth)
-     
-      // Move to the next relevant tag
-      seekToNextElement()
-      
-      while (xmlStreamReader.isStartElement && currentDepth >= 0) {
-        xmlStreamReader.require(START_ELEMENT, null, itemName)
-        val item = unmarshaller.unmarshal(xmlStreamReader, itemClass).getValue
-        f(item)
-        seekToNextElement()
-      }
-      
-      consumeRestOfStream(xmlStreamReader)
-    }
-  }
-  
-  /**
-   * Seeks to the next item element we are looking for
-   */
-  protected def seekToNextElementImpl(xmlStreamReader: XMLStreamReader2, _currentDepth: Int): Int = {
-    var currentDepth: Int = _currentDepth
-    var done: Boolean = false
 
-    while (!done) {
-      if (xmlStreamReader.isStartElement) {
-        // We found the parent or child element we are looking for
-        if(xmlStreamReader.getLocalName() == path(currentDepth)) {
-          
-          if (currentDepth == targetDepth) {
-            // This is the child element we are looking for
-            done = true
+      while (!done) {
+        var doNext: Boolean = true
+
+        if (xmlStreamReader.isStartElement) {
+          // Do we care about this path?
+          val pathMatches: Boolean = targets.exists{ t: XmlReaderPath[_, T] => t.pathMatches(xmlStreamReader, currentDepth, currentPath) }
```

Comment:
Might be better to avoid the closure here.

Reply:
What's the different between `pathMatches` and `matches`?  This code along with the branch that is followed when `pathMatches` is `true` is a little confusing and seems like it's checking the same thing multiple times.

Reply:
Good point, this got skewed pretty quickly, the matches should simply become  an tagMatches that assumes it would never get called unless the path matches.  Basically we need a logic for the case where "/items" matches, but we aren't at the "/items/item" path yet.  I had a guard for "/items/item" in the `matches` but it probably isn't needed here.